File: 1056-capacity-to-ship-packages-within-d-days/solution.py
class Solution:
    def func(self,weights,cap):
        load=0
        day=1
        for i in range(len(weights)):
            if(load+weights[i]>cap):
                day=day+1
                load=weights[i]
            else:
                load=load+weights[i]
        return day
    def shipWithinDays(self, weights: List[int], days: int) -> int:
        start=max(weights)
        end=sum(weights)
        while(start<=end):
            mid=start+(end-start)//2
            if(self.func(weights,mid)<=days):
                end=mid-1
            else:
                start=mid+1
        return start
        

File: 1056-capacity-to-ship-packages-within-d-days/solution.java
class Solution {

    public int daysNeeded(int[] weights, int capacity) {
        int days = 1;
        int load = 0;

        for (int w : weights) {
            if (load + w <= capacity) {
                load += w;
            } else {
                days++;
                load = w;
            }
        }

        return days;
    }

    public int shipWithinDays(int[] weights, int days) {

        int low = 0;
        int high = 0;

        for (int w : weights) {
            low = Math.max(low, w); // at least the heaviest package
            high += w;              // all packages in one day
        }

        int ans = high;

        while (low <= high) {

            int mid = low + (high - low) / 2;

            int requiredDays = daysNeeded(weights, mid);

            if (requiredDays <= days) {
                ans = mid;
                high = mid - 1; // try smaller capacity
            } else {
                low = mid + 1;  // need bigger capacity
            }
        }

        return ans;
    }
}
